File: app/BsDoc/bs/src/code.ml
(*  Copyright 26-Jul-2018 ºDeme
    GNU General Public License - V3 <http://www.gnu.org/licenses/>
*)

open Main
open Ui

let bottom = q "div" []
  It.(map (fun _ -> q "p" [Html "&nbsp;"][]) (range 1 28) |> to_list)

let up_arrow =
  q "div" [Style "position: fixed;bottom: 0px;right: 20px"][
    link (fun _ -> location_assign "#") |> Domo.add [img "up"]]

let scroll_to : string -> unit =
  [%raw {| function (s) { document.getElementById(s).scrollIntoView(); } |}]

let show' mpath fpath hyper pg =
  let cat s1 s2 = s1 ^ s2 in
  let numbers =
    It.(Txt.join
      "\n"
      (map
        (fun n ->
          let sn = string_of_int n in
          let len = String.length sn in
          reduce sn (fun seed  _ -> cat "&nbsp;" seed) (range 0 (4 - len))
        )
        (range 1 Txt.(split "\n" pg |> count))))
  in
  let w =
    q "div" [][
      q "table" [][
        q "tr" [][
          q "td" [Klass "frame"][
            q "a" [
                Att ("href", "?" ^ mpath ^ "@" ^ fpath);
                Html (fpath)
              ][]]]];
      q "table" [
          Att_i ("border", 0);
          Att ("width", "100%");
          Att_i ("cellspacing", 0)
        ][
        q "tr" [][
          q "td" [
              Klass "prel";
              Att ("width", "10px");
              Html ("<pre>" ^ numbers ^ "</pre>")
            ][];
          q "td" [
              Klass "prer";
              Html ("<pre>" ^ pg ^ "<pre>")
            ][]]];
      bottom;
      up_arrow
    ]

  in (
    Main.show mpath w;
    let _ = (It.get (qq "title")) |> Domo.set [Text ("#" ^ fpath)] in ();
    if hyper <> "hp:" && hyper <> "hp::" then scroll_to (hyper) else ()
  )

let show mpath fpath hyper =
  let rq = Json.(Client.wrq [
    "page", wstring "code";
    "mpath", wstring mpath;
    "fpath", wstring fpath;
    "hyper", wstring hyper;
  ])
  in
  Client.send (client ()) rq (fun rp ->
      match Client.rrp rp "page" (Json.ropt Json.rstring) with
      | None ->
        let rq = Json.(Client.wrq [
          "page", wstring "index";  (* Page index is reused *)
          "rq", wstring "setMenu";
          "option", wstring "@";
        ])
        in
        Client.send (client ()) rq (fun _ -> location_assign "?@")
      | Some pg ->
        let total_menu = mpath ^ "@" ^ fpath ^ "&" ^ hyper in
        let rq = Json.(Client.wrq [
          "page", wstring "index";  (* Page index is reused *)
          "rq", wstring "setMenu";
          "option", wstring total_menu;
        ])
        in
        Client.send (client ()) rq (fun _ -> show' mpath fpath hyper pg)
    )

File: app/JsDoc/js/src/Code.js
// Copyright 06-Sept-2018 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/** Code page */
// eslint-disable-next-line
import Main from "./Main.js";
import Ui from "./dmjs/Ui.js";
import It from "./dmjs/It.js";

const $ = Ui.$;

/** Code page. */
export default class Code {
  /**
   * @param {!Main} main Main page
   */
  constructor (main) {
    /**
     * @private
     * @type {!Main}
     */
    this._main = main;
  }

  /** @private */
  mkFoot () {
    return $("div")
      .adds([...It.range(35).map(() => $("div").html("&nbsp"))])
      .add($("div").style("position: fixed;bottom: 0px;right: 20px")
        .add(Ui.link(() => location.assign("#"))
          .add(Ui.img("up"))))
    ;
  }

  /** @private */
  show2 (html) {
    html = html.trim();
    const main = this._main;
    const model = main.model;
    const module = model.module;
    const link = `?${model.sel}@${module}`;
    const numbers =
      [...It.range(html.split("\n").length).map(i => String(i + 1))].join("\n");
    const pg = $("div")
      .add($("table").klass("frame")
        .add($("tr")
          .add($("td").html(`<a href="${link}">${module}`))))
      .add($("br"))
      .add($("table").style("width:100%;").att("cellspacing", 0)
        .add($("tr")
          .add($("td").klass("prel")
            .add($("div")
              .add($("pre").style("text-align:right;").html(numbers))))
          .add($("td").klass("prer")
            .add($("div")
              .add($("pre").html(html))))))
      .add(this.mkFoot())
    ;
    main.dom.show(pg);

    if (main.model.link !== "hp:") {
      document.getElementById(main.model.link).scrollIntoView();
    }
  }

  /**
   * @return {Promise}
   */
  async show () {
    const self = this;
    const main = self._main;
    const sel = main.model.sel;
    const module = main.model.module;
    const path = main.model.paths.find(p => p.id === sel);
    if (path === undefined || module === "") {
      main.go("@");
      return;
    }

    const rq = {
      "page": "code",
      "path": path.path + "/" + module + ".js",
      "link": main.model.link
    };
    const rp = await main.client.send(rq);
    const html = rp["html"];
    if (html === undefined) {
      main.go("@");
      return;
    }
    self.show2(html);
  }
}
